File: python/src/main/python/pygw/statistics/field/count_min_sketch_statistic.py
#
# Copyright (c) 2013-2020 Contributors to the Eclipse Foundation

#
# See the NOTICE file distributed with this work for additional information regarding copyright
# ownership. All rights reserved. This program and the accompanying materials are made available
# under the terms of the Apache License, Version 2.0 which accompanies this distribution and is
# available at http://www.apache.org/licenses/LICENSE-2.0.txt
# ===============================================================================================
from pygw.config import geowave_pkg
from ..statistic import FieldStatistic
from ..statistic_type import FieldStatisticType
from ...base import GeoWaveObject
from ...base.java_transformer import JavaTransformer


class CountMinSketchStatistic(FieldStatistic):
    """
    Maintains an estimate of how may of each attribute value occurs in a set of data.
    """
    STATS_TYPE = FieldStatisticType(geowave_pkg.core.store.statistics.field.CountMinSketchStatistic.STATS_TYPE)

    def __init__(
            self,
            type_name=None,
            field_name=None,
            error_factor=0.001,
            probability_of_correctness=0.98,
            java_ref=None):
        if java_ref is None:
            if type_name is None and field_name is None:
                java_ref = geowave_pkg.core.store.statistics.field.CountMinSketchStatistic()
            else:
                java_ref = geowave_pkg.core.store.statistics.field.CountMinSketchStatistic(
                    type_name, field_name, float(error_factor), float(probability_of_correctness))
        super().__init__(java_ref, CountMinSketchTransformer())

    def set_error_factor(self, error_factor):
        self._java_ref.setErrorFactor(float(error_factor))

    def get_error_factor(self):
        return self._java_ref.getErrorFactor()

    def set_probability_of_correctness(self, probability_of_correctness):
        self._java_ref.setProbabilityOfCorrectness(float(probability_of_correctness))

    def get_probability_of_correctness(self):
        return self._java_ref.getProbabilityOfCorrectness()


class CountMinSketchTransformer(JavaTransformer):
    def transform(self, j_object):
        return CountMinSketch(j_object)


class CountMinSketch(GeoWaveObject):
    def get_relative_error(self):
        return self._java_ref.getRelativeError()

    def get_confidence(self):
        return self._java_ref.getConfidence()

    def estimate_count(self, item):
        return self._java_ref.estimateCount(item)

File: core/store/src/main/java/org/locationtech/geowave/core/store/statistics/field/CountMinSketchStatistic.java
/**
 * Copyright (c) 2013-2022 Contributors to the Eclipse Foundation
 *
 * <p> See the NOTICE file distributed with this work for additional information regarding copyright
 * ownership. All rights reserved. This program and the accompanying materials are made available
 * under the terms of the Apache License, Version 2.0 which accompanies this distribution and is
 * available at http://www.apache.org/licenses/LICENSE-2.0.txt
 */
package org.locationtech.geowave.core.store.statistics.field;

import java.nio.ByteBuffer;
import org.locationtech.geowave.core.index.ByteArrayUtils;
import org.locationtech.geowave.core.index.Mergeable;
import org.locationtech.geowave.core.index.VarintUtils;
import org.locationtech.geowave.core.store.api.DataTypeAdapter;
import org.locationtech.geowave.core.store.api.FieldStatistic;
import org.locationtech.geowave.core.store.api.StatisticValue;
import org.locationtech.geowave.core.store.entities.GeoWaveRow;
import org.locationtech.geowave.core.store.statistics.StatisticsIngestCallback;
import com.beust.jcommander.Parameter;
import com.clearspring.analytics.stream.frequency.CountMinSketch;
import com.clearspring.analytics.stream.frequency.FrequencyMergeException;

/**
 * Maintains an estimate of how may of each attribute value occurs in a set of data.
 *
 * <p> Default values:
 *
 * <p> Error factor of 0.001 with probability 0.98 of retrieving a correct estimate. The Algorithm
 * does not under-state the estimate.
 */
public class CountMinSketchStatistic extends
    FieldStatistic<CountMinSketchStatistic.CountMinSketchValue> {
  public static final FieldStatisticType<CountMinSketchValue> STATS_TYPE =
      new FieldStatisticType<>("COUNT_MIN_SKETCH");

  @Parameter(names = "--errorFactor", description = "Error factor.")
  private double errorFactor = 0.001;

  @Parameter(
      names = "--probabilityOfCorrectness",
      description = "Probability of retrieving a correct estimate.")
  private double probabilityOfCorrectness = 0.98;

  public CountMinSketchStatistic() {
    super(STATS_TYPE);
  }

  public CountMinSketchStatistic(final String typeName, final String fieldName) {
    super(STATS_TYPE, typeName, fieldName);
  }

  public CountMinSketchStatistic(
      final String typeName,
      final String fieldName,
      final double errorFactor,
      final double probabilityOfCorrectness) {
    super(STATS_TYPE, typeName, fieldName);
    this.errorFactor = errorFactor;
    this.probabilityOfCorrectness = probabilityOfCorrectness;
  }

  public void setErrorFactor(final double errorFactor) {
    this.errorFactor = errorFactor;
  }

  public double getErrorFactor() {
    return this.errorFactor;
  }

  public void setProbabilityOfCorrectness(final double probabilityOfCorrectness) {
    this.probabilityOfCorrectness = probabilityOfCorrectness;
  }

  public double getProbabilityOfCorrectness() {
    return this.probabilityOfCorrectness;
  }

  @Override
  public String getDescription() {
    return "Maintains an estimate of how many of each attribute value occurs in a set of data.";
  }

  @Override
  public boolean isCompatibleWith(Class<?> fieldClass) {
    return true;
  }

  @Override
  public CountMinSketchValue createEmpty() {
    return new CountMinSketchValue(this);
  }

  @Override
  protected int byteLength() {
    return super.byteLength() + Double.BYTES * 2;
  }

  @Override
  protected void writeBytes(ByteBuffer buffer) {
    super.writeBytes(buffer);
    buffer.putDouble(errorFactor);
    buffer.putDouble(probabilityOfCorrectness);
  }

  @Override
  protected void readBytes(ByteBuffer buffer) {
    super.readBytes(buffer);
    errorFactor = buffer.getDouble();
    probabilityOfCorrectness = buffer.getDouble();
  }

  public static class CountMinSketchValue extends StatisticValue<CountMinSketch> implements
      StatisticsIngestCallback {

    private CountMinSketch sketch;

    public CountMinSketchValue() {
      super(null);
      sketch = null;
    }

    public CountMinSketchValue(final CountMinSketchStatistic statistic) {
      super(statistic);
      sketch =
          new CountMinSketch(statistic.errorFactor, statistic.probabilityOfCorrectness, 7364181);
    }

    public long totalSampleSize() {
      return sketch.size();
    }

    public long count(final String item) {
      return sketch.estimateCount(item);
    }

    @Override
    public void merge(Mergeable merge) {
      if (merge instanceof CountMinSketchValue) {
        try {
          sketch = CountMinSketch.merge(sketch, ((CountMinSketchValue) merge).sketch);
        } catch (final FrequencyMergeException e) {
          throw new RuntimeException("Unable to merge sketches", e);
        }
      }
    }

    @Override
    public <T> void entryIngested(DataTypeAdapter<T> adapter, T entry, GeoWaveRow... rows) {
      final Object o =
          adapter.getFieldValue(entry, ((CountMinSketchStatistic) statistic).getFieldName());
      if (o == null) {
        return;
      }
      sketch.add(o.toString(), 1);
    }

    @Override
    public CountMinSketch getValue() {
      return sketch;
    }

    @Override
    public byte[] toBinary() {
      final byte[] data = CountMinSketch.serialize(sketch);
      final ByteBuffer buffer =
          ByteBuffer.allocate(VarintUtils.unsignedIntByteLength(data.length) + data.length);
      VarintUtils.writeUnsignedInt(data.length, buffer);
      buffer.put(data);
      return buffer.array();
    }

    @Override
    public void fromBinary(byte[] bytes) {
      final ByteBuffer buffer = ByteBuffer.wrap(bytes);
      final byte[] data = ByteArrayUtils.safeRead(buffer, VarintUtils.readUnsignedInt(buffer));
      sketch = CountMinSketch.deserialize(data);
    }
  }
}
